### ISSUE TITLE
Missing DWARF entries for passed-by-reference structs

### ISSUE THREAD

**Issue by RReverser** [emscripten-core/emscripten#13192]
Take this example C:

```c
typedef struct A {
	int x;
} A;

typedef struct B {
	int x;
	float y;
} B;

typedef struct C {
	float y;
} C;

void accept_a_b_c(A a, B b, C c) {}

int main() {
	accept_a_b_c(
		(A) { .x = 10 },
		(B) { .x = 20, .y = 30 },
		(C) { .y = 40 }
	);
}
```

When compiled with debug info, the DWARF entry for `accept_a_b_c` looks like this:

```
0x00000026:   DW_TAG_subprogram
                DW_AT_low_pc    (0x00000007)
                DW_AT_high_pc   (0x0000002d)
                DW_AT_frame_base        (DW_OP_WASM_location 0x0 0x5, DW_OP_stack_value)
                DW_AT_name      ("accept_a_b_c")
                DW_AT_decl_file (".\temp.c")
                DW_AT_decl_line (14)
                DW_AT_prototyped        (true)
                DW_AT_external  (true)

0x0000003a:     DW_TAG_formal_parameter
                  DW_AT_location        (DW_OP_fbreg +8)
                  DW_AT_name    ("a")
                  DW_AT_decl_file       (".\temp.c")
                  DW_AT_decl_line       (14)
                  DW_AT_type    (0x00000076 "A")

0x00000048:     DW_TAG_formal_parameter
                  DW_AT_location        (DW_OP_fbreg +0)
                  DW_AT_name    ("c")
                  DW_AT_decl_file       (".\temp.c")
                  DW_AT_decl_line       (14)
                  DW_AT_type    (0x00000096 "C")

0x00000056:     NULL
```

As you can see, the entry for parameter `b` just got skipped. As a result, it's not visible in debuggers either.

This seems to be happening for any structs that are large enough to require passing by-reference instead of by-value. It also seems to be a Clang issue (I can reproduce with bare Clang), but opening here for tracking.

/cc @dschuff @bmeurer for visibility

**Comment by RReverser** [comment]
Fun fact / separate issue: building with optimisations makes parameter `b` reappear, but all params are missing `DW_AT_location` now, including `a` and `c` that had `DW_AT_location` above, so the debugger can't find them and print their contents.

**Comment by dschuff** [comment]
I think this is a duplicate of https://bugs.chromium.org/p/chromium/issues/detail?id=1153644 ?

**Comment by aardappel** [comment]
Possible fix in the works here: https://reviews.llvm.org/D94140

Actually, that fixes their missing location, not their absence if unused, I will check that next.

**Comment by RReverser** [comment]
> I think this is a duplicate of [bugs.chromium.org/p/chromium/issues/detail?id=1153644](https://bugs.chromium.org/p/chromium/issues/detail?id=1153644) ?

Oh yeah, I forgot I submitted that one on DevTools side.

It's not publicly visible though, and this is more of an Emscripten / Clang issue, so I guess doesn't hurt tracking here too?

**Comment by dschuff** [comment]
I agree that it's better to track them here than on the Chromium tracker.

**Comment by bmeurer** [comment]
Issue on the Chromium tracker is public now.

**Comment by aardappel** [comment]
Follow-up to encode values being "indirect": https://reviews.llvm.org/D94347

Also, the separate issue of these args not showing up when unused for now is a "nice to have" for later, because fixing it is going to be relatively complicated. As it turns out, being able to inspect unused args is closely tied to them being allocas in the current LLVM implementation, as unused args in registers/locals get culled at ISEL, and recovering this information during DWARF writing is non-trivial either. Both would require a fair bit of Wasm specific plumbing in non-Wasm LLVM code.

**Comment by RReverser** [comment]
@aardappel Hmm, I don't understand two parts: 1) why are unused args removed in non-optimized builds - it seems like an optimisation that should be disabled by default? 2) even stranger then, if unused args get culled early in the process, what makes the unused arg re-appear in the optimized build where DWARF entry is correctly generated?

**Comment by aardappel** [comment]
@RReverser
1) This is due to how we lower `byval` args in Wasm. In most targets in `-O0` these would be a callee pointer stored on the stack (pointing to the caller stack). In our case we pass them in locals (registers). LLVM is set up to cull those when unused.
2) Both in `-O0` and `-O2` we can have correct register/local debug information, assuming they're not culled.

**Comment by RReverser** [comment]
> Both in -O0 and -O2 we can have correct register/local debug information, assuming they're not culled.

Right, but the point here is that the snippet above actually produces entry for `b` in optimised mode but not in unoptimised one. Here's optimised output:

```
0x00000026:   DW_TAG_subprogram
                DW_AT_low_pc	(0x00000000)
                DW_AT_high_pc	(0x00000000)
                DW_AT_frame_base	(DW_OP_WASM_location 0x3 0x0, DW_OP_stack_value)
                DW_AT_GNU_all_call_sites	(true)
                DW_AT_name	("accept_a_b_c")
                DW_AT_decl_file	(".\..\temp.c")
                DW_AT_decl_line	(16)
                DW_AT_prototyped	(true)
                DW_AT_external	(true)

0x0000003d:     DW_TAG_formal_parameter
                  DW_AT_name	("a")
                  DW_AT_decl_file	(".\..\temp.c")
                  DW_AT_decl_line	(16)
                  DW_AT_type	(0x00000081 "A")

0x00000048:     DW_TAG_formal_parameter
                  DW_AT_name	("b")
                  DW_AT_decl_file	(".\..\temp.c")
                  DW_AT_decl_line	(16)
                  DW_AT_type	(0x000000c8 "B")

0x00000053:     DW_TAG_formal_parameter
                  DW_AT_name	("c")
                  DW_AT_decl_file	(".\..\temp.c")
                  DW_AT_decl_line	(16)
                  DW_AT_type	(0x000000a1 "C")
```

As you can see, unlike in an unoptimised build, `b` is produced. So my question is, what makes it "unculled" here.

**Comment by aardappel** [comment]
@RReverser from what I can tell, in `-O2`, variables that are optimized out get are still added to the list of arguments when possible (see final for loop in `DwarfDebug::collectEntityInfo`), which doesn't happen in `-O0` (presumably because it is asssumed nothing gets optimized out).  There is no location data associated with it though.

**Comment by RReverser** [comment]
> There is no location data associated with it though.

Hmm I wonder if it would be still a better middle-ground solution. As in, if we could do the same for `-O0`, then the variable would be at least visible in Scope but DevTools could show "optimized out" or something like that for its value - which would be less surprising than variable disappearing altogether.

**Comment by bmeurer** [comment]
Interesting information, thanks. We noticed that with WebAssembly target, `-O2` is almost unusable for debugging, since very very few variables have `DW_AT_location`s (see [this investigation](https://docs.google.com/document/d/1MI68Nn7G6CSTPFF7Rkn_Nh0q_Dy-1GtHdQhXUt7MM_o/edit#bookmark=id.qtbgzaote4be) (Googlers only)).

**Comment by dschuff** [comment]
Let's keep this open for now.
